Remove argument unpacking from variable kernels

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42043 Remove argument unpacking from variable kernels**
* #42025 IValue stores undefined tensor as None

This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.


Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

smessmer committed:
Remove argument unpacking from variable kernels

This is likely left over from when Variable was a different type than Tensor.
This is not needed anymore.

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove argument unpacking from variable kernels"


This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.


Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove argument unpacking from variable kernels"


This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.


Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove argument unpacking from variable kernels"


This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.


Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

[ghstack-poisoned]

bhosmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -201,10 +201,7 @@
-auto${ref} ${arg_name}_ = unpack${suffix}(${arg_name}, "${arg_name}", ${arg_pos});""")
-
-UNPACK_OPTIONS = CodeTemplate("""\
+PACK_TENSOR_OPTIONS = CodeTemplate("""\
```
Why pack? "unpack" is used pretty universally in these scripts to mean "process a param and assign the result to a local", which I think is still what this is doing, right?

smessmer replied:
that's how "unpack" was used previously but except for tensor options, that is now gone - so afaik, TensorOptions is the only arguments for which we use that nomenclature now. And to "unpack" TensorOptions in my mind means to take a TensorOptions object and scatter it into ScalarType, Layout, Device.

bhosmer replied:
Hmmm what about all this? https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_python_functions.py#L216

bhosmer reviewed (approved):
Nice! Accepting mod a) test failures? and b) inline nit 😁  (also full disclosure, if there's a risk of subtle ripple effects here then mine probably shouldn't be the final review)

ezyang reviewed (approved):
good riddance

smessmer committed:
Update on "Remove argument unpacking from variable kernels"


This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.


Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove argument unpacking from variable kernels"


This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.


Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove argument unpacking from variable kernels"


This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.


Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove argument unpacking from variable kernels"


This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.


Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove argument unpacking from variable kernels"


This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.


Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove argument unpacking from variable kernels"


This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.


Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove argument unpacking from variable kernels"


This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.


Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove argument unpacking from variable kernels"


This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.


Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove argument unpacking from variable kernels"


This is likely left over from when Variable was a different type than Tensor. At least after the recent c10::optional cleanups, this is now not needed anymore. Removing it simplifies our variable kernels.


Difference in generated code: P136668781

Differential Revision: [D22741968](https://our.internmc.facebook.com/intern/diff/D22741968/)

[ghstack-poisoned]
